### ISSUE TITLE
[BUG] SC2218 false negative for every code above last `"$(f)"`

### ISSUE THREAD

**Issue by loynoir** [koalaman/shellcheck#3290]
# bug

SC2218 false negative for every code above last `"$(f)"`

#### For bugs with existing features

- Rule Id (if any, e.g. SC1000): SC2218
- My shellcheck version (`shellcheck --version` or "online"):

```console
ShellCheck - shell script analysis tool
version: 0.11.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
```
- [ ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
set -euo pipefail

function pre_log() {
  echo "pre_log"
}

function pre_do_something() {
  echo pre_do_something
}

function post_log() {
  echo "post_log"
}

function post_do_something() {
  echo post_do_something
}

function main() {
  echo "[log] $(pre_log)"
  pre_do_something

  echo "[log] $(post_log)"
  post_do_something
}

main "$@"
```

#### Here's what shellcheck currently says:

```console
$ shellcheck /tmp/x3.sh && echo OK

In /tmp/x3.sh line 21:
  echo "[log] $(pre_log)"
                ^-----^ SC2218 (error): This function is only defined later. Move the definition up.


In /tmp/x3.sh line 22:
  pre_do_something
  ^--------------^ SC2218 (error): This function is only defined later. Move the definition up.

For more information:
  https://www.shellcheck.net/wiki/SC2218 -- This function is only defined lat...
```

Within simplify reproduce

- `"[log] $(post_log)"` is the last `"$(f)"`
- every code above last `"$(f)"`, are reported SC2218
  - `echo "[log] $(pre_log)"`
  - `pre_do_something`
- **but these SC2218 are false negative**
- every code not above last `"$(f)"`, are OK
  - `echo "[log] $(post_log)"`
  - `post_do_something`

#### Here's what I wanted or expected to see:

- every code above last `"$(f)"`, are **NOT** reported SC2218
  - `echo "[log] $(pre_log)"`
  - `pre_do_something`

**Comment by coiby** [comment]
I notice shellcheck v0.11 but not v0.10 also give false warning about [one of my shellspec tests](https://github.com/rhkdump/kdump-utils/blob/e22ca9248de44a71cdd51e0630456e53fe2554c1/spec/kdumpctl_setup_crypttab_spec.sh#L81).
```
In spec/kdumpctl_setup_crypttab_spec.sh line 81:
                                The stderr should include "Device UUID=$(get_all_kdump_crypt_dev) doesn't exist"
                                                                         ^---------------------^ SC2218 (error): This function is only defined later. Move the definition up

```
 
A simplified spec/kdumpctl_setup_crypttab_spec.sh can still reproduce this issue is as follows,
```bash
Describe "kdumpctl "
		Include ./kdumpctl

		Context "with safety checks and edge cases"

			It "aborts if target UUID is only in a commented-out line"
				get_all_kdump_crypt_dev() { echo "uuid-001"; }
				When call get_all_kdump_crypt_dev
				The status should be success
				The output should include "$(get_all_kdump_crypt_dev)"
			End

			It "succeeds with no changes if crypttab is already correct"
				get_all_kdump_crypt_dev() { echo "uuid-001"; }
				echo "luks-001 UUID=uuid-001 none link-volume-key=@u::%logon:${LUKS_KEY_PRFIX}uuid-001" >"$CRYPTTAB_FILE"
				When call setup_crypttab
				The status should be success
				The output should include "No changes were needed."
			End
		End
End
```

**Comment by etal** [comment]
This reproduces only on a ShellCheck built against fgl earlier than 5.8.1.0, and not on the released 0.11.0 binary, which is why it cannot be reproduced on shellcheck.net.

Measured on three osx-arm64 binaries, each reporting `version: 0.11.0`, each from the v0.11.0 source, with the snippet above:

| build | fgl | result |
| --- | --- | --- |
| `stack` with the tree's own `stack.yaml` (resolver lts-18.15) | 5.7.0.3 | SC2218 at 21:17 and 22:3, the same two lines quoted here |
| upstream's released `darwin.aarch64` | 5.8.3.0 | no SC2218 |
| `cabal` against a current Hackage index | 5.8.3.0 | no SC2218 |

Since 0.11.0, `checkUseBeforeDefinition` decides post-dominance through fgl's `dom`; in 0.10.0 the check was a syntactic reverse scan with no control-flow input, which is why 0.10.0 was clean on this code. Through fgl 5.8.0.0, `dom` reported every node of a graph as a dominator of any node unreachable from the root, so the post-dominance test is vacuously true for a call site that cannot reach the graph's exit. fgl 5.8.1.0 narrowed it to reachable nodes. Details and a five-line reproducer are in #3537.

To check which build you have:

```sh
strings -a "$(command -v shellcheck)" | grep -o 'fgl-[0-9][0-9.]*' | sort -u
```

`fgl-5.7.0.3` or `fgl-5.8.0.0` means the binary has this defect; `fgl-5.8.3.0` is what the releases carry. A downloaded release binary, or a `cabal` build with `--constraint='fgl >= 5.8.1.1'`, avoids it.
